File: backend/app/services/indicators.py
from app.models import Candle


def _rolling_midpoint(candles: list[Candle], period: int) -> list[float | None]:
    result: list[float | None] = [None] * len(candles)
    for i in range(period - 1, len(candles)):
        window = candles[i - period + 1 : i + 1]
        result[i] = (max(c.high for c in window) + min(c.low for c in window)) / 2
    return result


def ema(candles: list[Candle], period: int = 200) -> list[float]:
    alpha = 2 / (period + 1)
    output = [candles[0].close]
    for candle in candles[1:]:
        output.append(candle.close * alpha + output[-1] * (1 - alpha))
    return output

def ema_values(values: list[float], period: int) -> list[float | None]:
    """EMA on arbitrary float series."""
    if len(values) < period:
        return [None] * len(values)
    alpha = 2 / (period + 1)
    out: list[float | None] = [None] * len(values)
    # SMA seed
    sma = sum(values[:period]) / period
    out[period - 1] = sma
    prev = sma
    for i in range(period, len(values)):
        prev = values[i] * alpha + prev * (1 - alpha)
        out[i] = prev
    return out


def rsi(candles: list[Candle], period: int = 7) -> list[float | None]:
    output: list[float | None] = [None] * len(candles)
    if len(candles) <= period:
        return output
    gains, losses = [], []
    for i in range(1, period + 1):
        change = candles[i].close - candles[i - 1].close
        gains.append(max(change, 0)); losses.append(max(-change, 0))
    avg_gain, avg_loss = sum(gains) / period, sum(losses) / period
    output[period] = 100 if avg_loss == 0 else 100 - (100 / (1 + avg_gain / avg_loss))
    for i in range(period + 1, len(candles)):
        change = candles[i].close - candles[i - 1].close
        avg_gain = (avg_gain * (period - 1) + max(change, 0)) / period
        avg_loss = (avg_loss * (period - 1) + max(-change, 0)) / period
        output[i] = 100 if avg_loss == 0 else 100 - (100 / (1 + avg_gain / avg_loss))
    return output


def atr(candles: list[Candle], period: int = 14) -> list[float | None]:
    output: list[float | None] = [None] * len(candles)
    true_ranges = [candles[0].high - candles[0].low]
    for i in range(1, len(candles)):
        c, previous = candles[i], candles[i - 1]
        true_ranges.append(max(c.high - c.low, abs(c.high - previous.close), abs(c.low - previous.close)))
    if len(candles) < period:
        return output
    value = sum(true_ranges[:period]) / period
    output[period - 1] = value
    for i in range(period, len(candles)):
        value = (value * (period - 1) + true_ranges[i]) / period
        output[i] = value
    return output


def session_vwap(candles: list[Candle]) -> list[float]:
    output, cumulative_pv, cumulative_volume = [], 0.0, 0.0
    active_day = None
    for candle in candles:
        day = candle.timestamp.date()
        if day != active_day:
            active_day, cumulative_pv, cumulative_volume = day, 0.0, 0.0
        volume = max(candle.volume, 1)
        typical = (candle.high + candle.low + candle.close) / 3
        cumulative_pv += typical * volume
        cumulative_volume += volume
        output.append(cumulative_pv / cumulative_volume)
    return output


def ichimoku(candles: list[Candle], tenkan_period=7, kijun_period=22, span_b_period=44, displacement=22):
    tenkan = _rolling_midpoint(candles, tenkan_period)
    kijun = _rolling_midpoint(candles, kijun_period)
    raw_a: list[float | None] = [None] * len(candles)
    raw_b = _rolling_midpoint(candles, span_b_period)
    for i, (t, k) in enumerate(zip(tenkan, kijun)):
        if t is not None and k is not None:
            raw_a[i] = (t + k) / 2
    # Value projected in the past is the cloud visible at the current bar.
    cloud_a = [raw_a[i - displacement] if i >= displacement else None for i in range(len(candles))]
    cloud_b = [raw_b[i - displacement] if i >= displacement else None for i in range(len(candles))]
    return {"tenkan": tenkan, "kijun": kijun, "span_a": cloud_a, "span_b": cloud_b}


# ─── Complementary methods ──────────────────────────────────────────────

def macd(candles: list[Candle], fast=12, slow=26, signal=9):
    closes = [c.close for c in candles]
    ema_fast = ema_values(closes, fast)
    ema_slow = ema_values(closes, slow)
    macd_line: list[float | None] = [None] * len(candles)
    for i in range(len(candles)):
        if ema_fast[i] is not None and ema_slow[i] is not None:
            macd_line[i] = float(ema_fast[i] - ema_slow[i])  # type: ignore
    # signal line = EMA of macd_line
    valid = [(i, v) for i, v in enumerate(macd_line) if v is not None]
    vals = [v for _, v in valid]
    sig = ema_values(vals, signal) if len(vals) >= signal else [None]*len(vals)
    # map back
    signal_line: list[float | None] = [None] * len(candles)
    hist: list[float | None] = [None] * len(candles)
    for idx, (orig_i, _) in enumerate(valid):
        if sig[idx] is not None:
            signal_line[orig_i] = sig[idx]
            hist[orig_i] = float(macd_line[orig_i] - sig[idx])  # type: ignore
    return {"macd": macd_line, "signal": signal_line, "histogram": hist}


def bollinger(candles: list[Candle], period=20, mult=2.0):
    closes = [c.close for c in candles]
    sma: list[float | None] = [None]*len(candles)
    upper: list[float | None] = [None]*len(candles)
    lower: list[float | None] = [None]*len(candles)
    for i in range(period-1, len(candles)):
        window = closes[i-period+1:i+1]
        mean = sum(window)/period
        var = sum((x-mean)**2 for x in window)/period
        std = var ** 0.5
        sma[i]=mean
        upper[i]=mean+mult*std
        lower[i]=mean-mult*std
    return {"middle": sma, "upper": upper, "lower": lower, "width": [ (u-l) if u and l else None for u,l in zip(upper, lower)] }


def stochastic(candles: list[Candle], k_period=14, d_period=3):
    k: list[float | None] = [None]*len(candles)
    for i in range(k_period-1, len(candles)):
        window = candles[i-k_period+1:i+1]
        hh = max(c.high for c in window)
        ll = min(c.low for c in window)
        denom = hh - ll
        if denom == 0:
            k[i]=50
        else:
            k[i]= 100*(candles[i].close - ll)/denom
    d: list[float | None] = [None]*len(candles)
    # %D = SMA of %K
    for i in range(len(candles)):
        if i >= k_period-1 + d_period-1:
            vals = [v for v in k[i-d_period+1:i+1] if v is not None]
            if len(vals)==d_period:
                d[i]= sum(vals)/d_period
    return {"k": k, "d": d}


def adx(candles: list[Candle], period=14):
    """Simplified ADX for trend strength."""
    n = len(candles)
    out: list[float | None] = [None]*n
    if n < period*2:
        return out
    tr = [0.0]*n
    plus_dm = [0.0]*n
    minus_dm = [0.0]*n
    for i in range(1, n):
        h, l, pc = candles[i].high, candles[i].low, candles[i-1].close
        ph, pl = candles[i-1].high, candles[i-1].low
        tr[i]= max(h-l, abs(h-pc), abs(l-pc))
        up = h - ph
        dn = pl - l
        if up > dn and up > 0:
            plus_dm[i]= up
        if dn > up and dn > 0:
            minus_dm[i]= dn
    # Wilder smoothing
    atr_s = sum(tr[1:period+1])/period
    pdm_s = sum(plus_dm[1:period+1])/period
    mdm_s = sum(minus_dm[1:period+1])/period
    dx: list[float | None] = [None]*n
    for i in range(period, n):
        if i > period:
            atr_s = (atr_s*(period-1)+ tr[i])/period
            pdm_s = (pdm_s*(period-1)+ plus_dm[i])/period
            mdm_s = (mdm_s*(period-1)+ minus_dm[i])/period
        pdi = 100*pdm_s/atr_s if atr_s else 0
        mdi = 100*mdm_s/atr_s if atr_s else 0
        denom = pdi+mdi
        dx[i]= 100*abs(pdi-mdi)/denom if denom else 0
    # ADX = Wilder MA of DX
    adx_val = sum(v for v in dx[period:period*2] if v is not None)/period
    out[period*2-1]= adx_val
    for i in range(period*2, n):
        if dx[i] is not None and out[i-1] is not None:
            out[i]= (out[i-1]*(period-1)+ dx[i])/period # type: ignore
    return out


def support_resistance(candles: list[Candle], lookback=30):
    """Simple recent swing levels."""
    if len(candles) < lookback:
        return {"support": None, "resistance": None}
    recent = candles[-lookback:]
    return {"support": min(c.low for c in recent), "resistance": max(c.high for c in recent)}

File: android/app/src/main/java/com/aurum/edge/engine/Indicators.kt
package com.aurum.edge.engine

import com.aurum.edge.core.Candle
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

/**
 * Pure indicator math over *real* closed candles.
 * Every function returns a list aligned with the input (null where there is not enough history),
 * so a value at index i depends only on bars 0..i — no look-ahead.
 */
object Indicators {

    fun donchianMid(candles: List<Candle>, period: Int): List<Double?> {
        if (period <= 0) return List(candles.size) { null }
        val out = MutableList<Double?>(candles.size) { null }
        for (i in candles.indices) {
            if (i < period - 1) continue
            var hi = Double.NEGATIVE_INFINITY
            var lo = Double.POSITIVE_INFINITY
            for (j in i - period + 1..i) {
                hi = max(hi, candles[j].high)
                lo = min(lo, candles[j].low)
            }
            out[i] = (hi + lo) / 2.0
        }
        return out
    }

    fun ema(values: List<Double>, period: Int): List<Double?> {
        if (values.isEmpty() || period <= 1) return List(values.size) { null }
        val out = MutableList<Double?>(values.size) { null }
        if (values.size < period) return out
        val k = 2.0 / (period + 1)
        var prev = values.take(period).average()
        out[period - 1] = prev
        for (i in period until values.size) {
            prev = values[i] * k + prev * (1 - k)
            out[i] = prev
        }
        return out
    }

    fun rsi(candles: List<Candle>, period: Int = 7): List<Double?> {
        val out = MutableList<Double?>(candles.size) { null }
        if (candles.size <= period) return out
        var gains = 0.0
        var losses = 0.0
        for (i in 1..period) {
            val d = candles[i].close - candles[i - 1].close
            if (d > 0) gains += d else losses -= d
        }
        var avgGain = gains / period
        var avgLoss = losses / period
        out[period] = if (avgLoss == 0.0) 100.0 else 100.0 - 100.0 / (1 + avgGain / avgLoss)
        for (i in period + 1 until candles.size) {
            val d = candles[i].close - candles[i - 1].close
            avgGain = (avgGain * (period - 1) + max(d, 0.0)) / period
            avgLoss = (avgLoss * (period - 1) + max(-d, 0.0)) / period
            out[i] = if (avgLoss == 0.0) 100.0 else 100.0 - 100.0 / (1 + avgGain / avgLoss)
        }
        return out
    }

    fun atr(candles: List<Candle>, period: Int = 14): List<Double?> {
        val out = MutableList<Double?>(candles.size) { null }
        if (candles.size <= period) return out
        val tr = DoubleArray(candles.size)
        for (i in candles.indices) {
            tr[i] = if (i == 0) {
                candles[0].high - candles[0].low
            } else {
                max(
                    candles[i].high - candles[i].low,
                    max(abs(candles[i].high - candles[i - 1].close), abs(candles[i].low - candles[i - 1].close)),
                )
            }
        }
        var value = tr.take(period).average()
        out[period - 1] = value
        for (i in period until candles.size) {
            value = (value * (period - 1) + tr[i]) / period
            out[i] = value
        }
        return out
    }

    fun macdHistogram(candles: List<Candle>, fast: Int = 12, slow: Int = 26, signal: Int = 9): List<Double?> {
        val closes = candles.map { it.close }
        val emaFast = ema(closes, fast)
        val emaSlow = ema(closes, slow)
        val line = closes.indices.map { i ->
            val f = emaFast[i]
            val s = emaSlow[i]
            if (f == null || s == null) null else f - s
        }
        val compact = line.filterNotNull()
        val signalEma = ema(compact, signal)
        val out = MutableList<Double?>(candles.size) { null }
        var idx = 0
        for (i in line.indices) {
            val value = line[i] ?: continue
            val sig = signalEma.getOrNull(idx)
            if (sig != null) out[i] = value - sig
            idx++
        }
        return out
    }

    fun bollinger(candles: List<Candle>, period: Int = 20, mult: Double = 2.0): Triple<List<Double?>, List<Double?>, List<Double?>> {
        val upper = MutableList<Double?>(candles.size) { null }
        val lower = MutableList<Double?>(candles.size) { null }
        val mid = MutableList<Double?>(candles.size) { null }
        for (i in candles.indices) {
            if (i < period - 1) continue
            val window = candles.subList(i - period + 1, i + 1).map { it.close }
            val mean = window.average()
            val sd = sqrt(window.sumOf { (it - mean) * (it - mean) } / period)
            mid[i] = mean
            upper[i] = mean + mult * sd
            lower[i] = mean - mult * sd
        }
        return Triple(upper, mid, lower)
    }

    fun adx(candles: List<Candle>, period: Int = 14): List<Double?> {
        val out = MutableList<Double?>(candles.size) { null }
        if (candles.size <= period * 2) return out
        val plusDm = DoubleArray(candles.size)
        val minusDm = DoubleArray(candles.size)
        val tr = DoubleArray(candles.size)
        for (i in 1 until candles.size) {
            val up = candles[i].high - candles[i - 1].high
            val down = candles[i - 1].low - candles[i].low
            plusDm[i] = if (up > down && up > 0) up else 0.0
            minusDm[i] = if (down > up && down > 0) down else 0.0
            tr[i] = max(
                candles[i].high - candles[i].low,
                max(abs(candles[i].high - candles[i - 1].close), abs(candles[i].low - candles[i - 1].close)),
            )
        }
        var trSum = tr.drop(1).take(period).sum()
        var plusSum = plusDm.drop(1).take(period).sum()
        var minusSum = minusDm.drop(1).take(period).sum()
        val dx = DoubleArray(candles.size)
        for (i in period + 1 until candles.size) {
            trSum = trSum - trSum / period + tr[i]
            plusSum = plusSum - plusSum / period + plusDm[i]
            minusSum = minusSum - minusSum / period + minusDm[i]
            if (trSum <= 0.0) continue
            val plusDi = 100.0 * plusSum / trSum
            val minusDi = 100.0 * minusSum / trSum
            val sum = plusDi + minusDi
            dx[i] = if (sum == 0.0) 0.0 else 100.0 * abs(plusDi - minusDi) / sum
        }
        var adxValue: Double? = null
        val firstAdxIndex = period * 2
        if (firstAdxIndex < candles.size) {
            adxValue = dx.drop(period + 1).take(period).average()
            out[firstAdxIndex] = adxValue
            for (i in firstAdxIndex + 1 until candles.size) {
                adxValue = ((adxValue ?: 0.0) * (period - 1) + dx[i]) / period
                out[i] = adxValue
            }
        }
        return out
    }

    /** Session VWAP, reset at 00:00 UTC. Falls back to a plain typical-price average when the feed reports no volume. */
    fun sessionVwap(candles: List<Candle>, sessionMillis: Long = 86_400_000L): List<Double> {
        val out = MutableList(candles.size) { 0.0 }
        var pv = 0.0
        var vol = 0.0
        var priceSum = 0.0
        var count = 0
        var sessionStart = Long.MIN_VALUE
        for (i in candles.indices) {
            val c = candles[i]
            val session = c.time / sessionMillis
            if (session != sessionStart) {
                sessionStart = session
                pv = 0.0; vol = 0.0; priceSum = 0.0; count = 0
            }
            val typical = (c.high + c.low + c.close) / 3.0
            pv += typical * c.volume
            vol += c.volume
            priceSum += typical
            count++
            out[i] = when {
                vol > 0.0 -> pv / vol
                count > 0 -> priceSum / count
                else -> c.close
            }
        }
        return out
    }

    fun relativeVolume(candles: List<Candle>, period: Int = 30): List<Double?> {
        val out = MutableList<Double?>(candles.size) { null }
        for (i in candles.indices) {
            if (i < period) continue
            val avg = candles.subList(i - period, i).map { it.volume }.average()
            out[i] = if (avg <= 0.0) null else candles[i].volume / avg
        }
        return out
    }
}

/**
 * Ichimoku 7/22/44 (1m) or 9/26/52 (5m+).
 * `senkouA/B` are stored at their computation index; the execution-safe cloud for bar i
 * is read from `cloudAt(i)`, which uses the values computed `displacement` bars earlier.
 */
data class Ichimoku(
    val tenkan: List<Double?>,
    val kijun: List<Double?>,
    val senkouA: List<Double?>,
    val senkouB: List<Double?>,
    val displacement: Int,
) {
    /** Cloud as it is actually visible at bar [index] (no look-ahead). */
    fun cloudAt(index: Int): Pair<Double, Double>? {
        val i = index - displacement
        if (i < 0) return null
        val a = senkouA.getOrNull(i) ?: return null
        val b = senkouB.getOrNull(i) ?: return null
        return a to b
    }

    companion object {
        fun compute(candles: List<Candle>, tenkanPeriod: Int, kijunPeriod: Int, spanBPeriod: Int, displacement: Int): Ichimoku {
            val tenkan = Indicators.donchianMid(candles, tenkanPeriod)
            val kijun = Indicators.donchianMid(candles, kijunPeriod)
            val spanB = Indicators.donchianMid(candles, spanBPeriod)
            val spanA = MutableList<Double?>(candles.size) { null }
            for (i in candles.indices) {
                val t = tenkan[i]
                val k = kijun[i]
                if (t != null && k != null) spanA[i] = (t + k) / 2.0
            }
            return Ichimoku(tenkan, kijun, spanA, spanB, displacement)
        }
    }
}
